src/shogun/util/hash.h

```diff
@@ -0,0 +1,37 @@
+/*
+ * This software is distributed under BSD 3-clause license (see LICENSE file).
+ *
+ * Authors: Viktor Gal
+ */
+
+#ifndef __UTIL_HASH_H__
+#define __UTIL_HASH_H__
+
+namespace shogun
+{
+	class CSGObject;
+}
+
+namespace std
+{
+	template <typename T, enable_if_t<is_base_of_v<shogun::CSGObject, T>>* = nullptr>
+	using sgo_base = T;
+
+	template<typename T> struct hash<sgo_base<T>>
+	{
+		typedef T argument_type;
+		typedef std::size_t result_type;
+
+		result_type operator()(const argument_type& o) const
+		{
+			auto params = o.get_params();
+			result_type seed = 0;
+			for (const auto& p: params)
+				seed ^= p.second->get_value().hash() + 0x9e3779b9 + (seed<<6) + (seed>>2);
```

Comment:
Are the numbers here arbitrary  or something standard?

Reply:
https://www.boost.org/doc/libs/1_55_0/doc/html/hash/combine.html

Reply:
I see